crates/etc-merge/src/lib.rs

```diff
@@ -0,0 +1,1079 @@
+                }
+
+                Err(e) => Err(e).context(format!("{file_name:?}"))?,
+            },
+        }
+
+        current_path.pop();
+    }
+
+    Ok(())
+}
+
+// 1. Files in the currently booted deployment’s /etc which were modified from the default /usr/etc (of the same deployment) are retained.
+//
+// 2. Files in the currently booted deployment’s /etc which were not modified from the default /usr/etc (of the same deployment)
+// are upgraded to the new defaults from the new deployment’s /usr/etc.
+
+// Modifications
+// 1. File deleted from new /etc
+// 2. File added in new /etc
+//
+// 3. File modified in new /etc
+//    a. Content added/deleted
+//    b. Permissions/ownership changed
+//    c. Was a file but changed to directory/symlink etc or vice versa
+//    d. xattrs changed - we don't include this right now
+#[context("Getting modifications")]
+fn get_modifications(
+    pristine: &Directory<CustomMetadata>,
+    current: &Directory<CustomMetadata>,
+    mut current_path: PathBuf,
+    diff: &mut Diff,
+) -> anyhow::Result<()> {
+    use composefs::generic_tree::LeafContent::*;
+
+    for (path, inode) in current.sorted_entries() {
+        current_path.push(path);
+
+        match inode {
+            Inode::Directory(curr_dir) => {
+                match pristine.get_directory(path) {
+                    Ok(old_dir) => {
+                        if !stat_eq_ignore_mtime(&curr_dir.stat, &old_dir.stat) {
+                            // Directory permissions/owner modified
+                            diff.modified.push(current_path.clone());
+                        }
+
+                        get_modifications(old_dir, &curr_dir, current_path.clone(), diff)?
+                    }
+
+                    Err(ImageError::NotFound(..)) => {
+                        // Dir not found in original /etc, dir was added
+                        diff.added.push(current_path.clone());
+
+                        // Also add every file inside that dir
+                        diff.added
+                            .extend(collect_all_files(&curr_dir, current_path.clone()));
+                    }
+
+                    Err(ImageError::NotADirectory(..)) => {
+                        // Some directory was changed to a file/symlink
+                        // This should be counted in the diff, but we don't really merge this
+                        diff.modified.push(current_path.clone());
+                    }
+
+                    Err(e) => Err(e)?,
+                }
+            }
+
+            Inode::Leaf(leaf) => match pristine.ref_leaf(path) {
+                Ok(old_leaf) => {
+                    match (&old_leaf.content, &leaf.content) {
+                        (Regular(old_meta), Regular(current_meta)) => {
+                            if old_meta.content_hash != current_meta.content_hash
+                                || !stat_eq_ignore_mtime(&old_leaf.stat, &leaf.stat)
```

Comment:
I think we could move this conditional to line 228 so it's not repeated for the regfile and symlink cases